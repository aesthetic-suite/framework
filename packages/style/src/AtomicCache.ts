import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
