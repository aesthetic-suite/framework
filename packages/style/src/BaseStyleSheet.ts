import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
