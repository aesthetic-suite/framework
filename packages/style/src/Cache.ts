import { Value, ValueWithFallbacks } from '@aesthetic/types';
import { CacheManager, CacheItem, CacheStorage, RenderOptions } from './types';

export function createCacheKey(
  property: string,
  value: Value | ValueWithFallbacks,
  { selector, conditions }: RenderOptions,
): string {
  let key = `${property}:${value}`;

  if (selector) {
    key += `:${selector}`;
  }

  if (conditions && conditions.length > 0) {
    key += `:${conditions.join(':')}`;
  }

  return key;
}

export function createCacheManager(defaultItems: CacheStorage = {}): CacheManager {
  const cache: CacheStorage = defaultItems;

  return {
    read(key, minimumRank) {
      const items = cache[key];

      if (!items) {
        return null;
      } else if (minimumRank === undefined) {
        return items[0];
      }

      return items.find((item) => item.rank! >= minimumRank) || null;
    },

    write(key: string, item: CacheItem) {
      const result = cache[key] || [];

      result.push(item);

      cache[key] = result;
    },
  };
}
