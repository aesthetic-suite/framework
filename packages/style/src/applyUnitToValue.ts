import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
