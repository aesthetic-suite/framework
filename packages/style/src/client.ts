import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import createStyleElements, { getStyleElement } from './client/dom';
import hydrateStyles from './client/hydration';
// Rollup compatibility
import {
  createCacheManager,
  createStyleEngine,
  createSheetManager,
  formatVariable,
  EngineOptions,
  StyleEngine,
} from './index';

export { createStyleElements, getStyleElement };

function setRootVariables(vars: Variables) {
  const root = document.documentElement;

  objectLoop(vars, (value, key) => {
    root.style.setProperty(formatVariable(key), String(value));
  });
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: createSheetManager(createStyleElements()),
      ...options,
    }),
    setRootVariables,
  };

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
