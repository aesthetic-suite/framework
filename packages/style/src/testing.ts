/* eslint-disable unicorn/import-index */

import { CSS } from '@aesthetic/types';
import {
  CacheManager,
  StyleEngine,
  EngineOptions,
  SheetManager,
  SheetType,
  StyleRule,
  getDocumentStyleSheet,
} from './index';
import { createCacheManager, createStyleElements, createEngine, createSheetManager } from './next';

export function createTestCacheManager(): CacheManager {
  return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
  return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(options: Partial<EngineOptions>): StyleEngine {
  return createEngine({
    cacheManager: createTestCacheManager(),
    sheetManager: createTestSheetManager(),
    ...options,
  });
}

export function getRenderedStyles(type: SheetType | StyleRule): CSS {
  return Array.from(
    ((typeof type === 'string' ? getDocumentStyleSheet(type) : type) as StyleRule).cssRules,
  ).reduce((css, rule) => css + rule.cssText, '');
}

export function purgeStyles(type?: SheetType): void {
  if (type) {
    // This is the only way to generate accurate snapshots.
    // It may slow down tests though?
    document.getElementById(`aesthetic-${type}`)?.remove();
  } else {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  }
}
