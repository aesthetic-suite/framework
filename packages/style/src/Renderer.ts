import hash from 'string-hash';
import { isObject } from 'aesthetic-utils';
import { hyphenateProperty } from 'css-in-js-utils';
import AtomicCache from './AtomicCache';
import { ClassName, Properties, StyleParams, Property, Value } from './types';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isMediaQueryCondition from './isMediaQueryCondition';
import isSupportsCondition from './isSupportsCondition';
import isNestedSelector from './isNestedSelector';

export default class Renderer {
  protected atRuleCache = new Map();

  protected classNameCache = new AtomicCache();

  constructor() {
    // Create the elements in a strict order
    getDocumentStyleSheet('global');
    getDocumentStyleSheet('low-pri');
    getDocumentStyleSheet('high-pri');
  }

  render(properties: Properties, params: StyleParams): ClassName {
    return Object.entries(properties)
      .map(([key, value]) => {
        if (value === null || value === undefined) {
          return '';
        }

        if (isObject(value)) {
          if (isMediaQueryCondition(key) || isSupportsCondition(key)) {
            return this.render(value, { ...params, condition: key });
          } else if (isNestedSelector(key)) {
            return this.render(value, { ...params, selector: key });
          } else if (__DEV__) {
            console.warn(`Unknown property "${key}" with value "${value}".`);
          }

          return '';
        }

        return this.renderDeclaration(key as Property, value, params);
      })
      .join(' ');
  }

  renderDeclaration<K extends Property>(property: K, value: Properties[K], params: StyleParams) {
    const key = hyphenateProperty(property);
    const item = this.classNameCache.read(key, value, params);

    if (item) {
      return item.className;
    }

    const sheet = getDocumentStyleSheet(params.type);
    const rank = sheet.cssRules.length;
    const className = this.createUniqueHash(property, value, params);

    sheet.insertRule(this.createRule(className, key, value, params), rank);

    this.classNameCache.write(key, value, {
      condition: '',
      selector: '',
      type: 'low-pri',
      ...params,
      className,
      rank,
    });

    return className;
  }

  renderFontFace() {}

  renderKeyframes() {}

  protected createDeclaration(property: string, value: Value): string {
    return `${property}: ${value}`;
  }

  protected createDeclarationBlock(properties: Properties): string {
    return Object.entries(properties)
      .map(([prop, value]) => this.createDeclaration(prop, value))
      .join(';');
  }

  protected createRule(
    className: ClassName,
    property: string,
    value: Value,
    params: StyleParams,
  ): string {
    let rule = `.${className}`;

    if (params.selector) {
      rule += params.selector;
    }

    rule += `{ ${this.createDeclaration(property, value)} }`;

    if (params.condition) {
      rule = `${params.condition} { ${rule} }`;
    }

    return rule;
  }

  protected createUniqueHash(property: string, value: Value, params: StyleParams): string {
    return hash(`${params.condition || ''}${params.selector || ''}${property}${value}`);
  }
}
