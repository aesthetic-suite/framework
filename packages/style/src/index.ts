/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
