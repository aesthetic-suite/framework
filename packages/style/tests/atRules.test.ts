import Renderer from '../src/client/ClientRenderer';
import { getRenderedStyles, purgeStyles } from '../src/testing';

describe('At-rules', () => {
  let renderer: Renderer;

  beforeEach(() => {
    renderer = new Renderer();
  });

  afterEach(() => {
    purgeStyles();
  });

  it('doesnt insert the same at-rule more than once', () => {
    renderer.renderFontFace({
      fontFamily: '"Open Sans"',
      fontStyle: 'normal',
      fontWeight: 800,
      src: 'url("fonts/OpenSans-Bold.woff2")',
    });
    renderer.renderFontFace({
      fontFamily: '"Open Sans"',
      fontStyle: 'normal',
      fontWeight: 800,
      src: 'url("fonts/OpenSans-Bold.woff2")',
    });
    renderer.renderKeyframes({
      from: {
        transform: 'translateX(0%)',
      },
      to: {
        transform: 'translateX(100%)',
      },
    });
    renderer.renderKeyframes({
      from: {
        transform: 'translateX(0%)',
      },
      to: {
        transform: 'translateX(100%)',
      },
    });

    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  it('inserts at-rules before standard rules', () => {
    renderer.renderRule({ display: 'block' }, { type: 'global' });

    renderer.renderFontFace({
      fontFamily: '"Open Sans"',
      fontStyle: 'normal',
      fontWeight: 800,
      src: 'url("fonts/OpenSans-Bold.woff2")',
    });

    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  describe('@font-face', () => {
    it('renders and returns family name', () => {
      const name = renderer.renderFontFace({
        fontFamily: '"Open Sans"',
        fontStyle: 'normal',
        fontWeight: 800,
        src: 'url("fonts/OpenSans-Bold.woff2")',
      });

      expect(name).toBe('"Open Sans"');
      expect(getRenderedStyles('global')).toMatchSnapshot();
    });

    it('errors if no family name provided', () => {
      expect(() => {
        renderer.renderFontFace({
          fontFamily: '',
          fontStyle: 'normal',
        });
      }).toThrow('Font faces require a font family.');
    });
  });

  describe('@import', () => {
    it('renders all variants', () => {
      renderer.renderImport('url("print.css") print');
      renderer.renderImport('url("a11y.css") speech');
      renderer.renderImport("'custom.css'");
      renderer.renderImport('url("chrome://communicator/skin")');
      renderer.renderImport('"common.css" screen;'); // Ends in semicolon
      renderer.renderImport("url('landscape.css') screen and (orientation: landscape)");

      expect(getRenderedStyles('global')).toMatchSnapshot();
    });
  });

  describe('@keyframes', () => {
    it('renders range based and returns animation name', () => {
      const name = renderer.renderKeyframes({
        from: {
          transform: 'translateX(0%)',
        },
        to: {
          transform: 'translateX(100%)',
        },
      });

      expect(name).toBe('kf103rcyx');
      expect(getRenderedStyles('global')).toMatchSnapshot();
    });

    it('renders percentage based and returns animation name', () => {
      const name = renderer.renderKeyframes({
        '0%': { top: 0, left: 0 },
        '30%': { top: '50px' },
        '68%, 72%': { left: '50px' },
        '100%': { top: '100px', left: '100%' },
      });

      expect(name).toBe('kf22exw8');
      expect(getRenderedStyles('global')).toMatchSnapshot();
    });

    it('can provide a custom animation name', () => {
      const name = renderer.renderKeyframes(
        {
          from: {
            opacity: 0,
          },
          to: {
            opacity: 1,
          },
        },
        'fade',
      );

      expect(name).toBe('fade');
      expect(getRenderedStyles('global')).toMatchSnapshot();
    });

    it('supports ltr and rtl', () => {
      const ltr = renderer.renderKeyframes({
        from: {
          left: '0',
        },
        to: {
          right: '100px',
        },
      });

      const rtl = renderer.renderKeyframes(
        {
          from: {
            left: '0',
          },
          to: {
            right: '100px',
          },
        },
        '',
        {
          rtl: true,
        },
      );

      expect(ltr).toBe('kf1lt4056');
      expect(rtl).toBe('kf944ipm');
      expect(getRenderedStyles('global')).toMatchSnapshot();
    });
  });
});
