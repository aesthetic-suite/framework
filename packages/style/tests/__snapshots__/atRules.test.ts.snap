// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`At-rules @font-face renders and returns family name 1`] = `"@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}"`;

exports[`At-rules @import renders all variants 1`] = `"@import url(print.css) print;@import url(a11y.css) speech;@import url(custom.css);@import url(chrome://communicator/skin);@import url(common.css) screen;@import url(landscape.css) screen and (orientation: landscape);"`;

exports[`At-rules @keyframes can provide a custom animation name 1`] = `
"@keyframes fade { 
  from {opacity: 0;} 
  to {opacity: 1;} 
}"
`;

exports[`At-rules @keyframes renders percentage based and returns animation name 1`] = `
"@keyframes kf22exw8 { 
  0% {top: 0; left: 0;} 
  30% {top: 50px;} 
  68%,72% {left: 50px;} 
  100% {top: 100px; left: 100%;} 
}"
`;

exports[`At-rules @keyframes renders range based and returns animation name 1`] = `
"@keyframes kf103rcyx { 
  from {transform: translateX(0%);} 
  to {transform: translateX(100%);} 
}"
`;

exports[`At-rules @keyframes supports ltr and rtl 1`] = `
"@keyframes kf1lt4056 { 
  from {left: 0;} 
  to {right: 100px;} 
}@keyframes kf944ipm { 
  from {right: 0;} 
  to {left: 100px;} 
}"
`;

exports[`At-rules doesnt insert the same at-rule more than once 1`] = `
"@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}@keyframes kf103rcyx { 
  from {transform: translateX(0%);} 
  to {transform: translateX(100%);} 
}"
`;

exports[`At-rules inserts at-rules before standard rules 1`] = `"@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}.a {display: block;}"`;
